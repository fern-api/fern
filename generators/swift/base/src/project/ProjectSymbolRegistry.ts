import { assertDefined, SymbolRegistry, values } from "@fern-api/core-utils";
import { swift } from "@fern-api/swift-codegen";

const SYMBOL_ID_PREFIX = "symbol_id:";

export class ProjectSymbolRegistry {
    private static readonly reservedSymbols = [
        "Swift",
        "Foundation",
        ...swift.Type.primitiveSymbolNames(),
        ...swift.Type.foundationSymbolNames(),
        ...values(swift.Protocol)
    ];

    public static create(): ProjectSymbolRegistry {
        return new ProjectSymbolRegistry(ProjectSymbolRegistry.reservedSymbols);
    }

    private readonly registry: SymbolRegistry;
    private readonly requestsRegistry: SymbolRegistry;

    private constructor(reservedSymbolNames: string[]) {
        this.registry = new SymbolRegistry({
            reservedSymbolNames: reservedSymbolNames
        });
        this.requestsRegistry = new SymbolRegistry({ reservedSymbolNames: [] });
    }

    public getModuleSymbolOrThrow(): string {
        const symbolName = this.registry.getSymbolNameById(this.getModuleSymbolId());
        assertDefined(symbolName, "Module symbol not found.");
        return symbolName;
    }

    public getRootClientSymbolOrThrow(): string {
        const symbolName = this.registry.getSymbolNameById(this.getRootClientSymbolId());
        assertDefined(symbolName, "Root client symbol not found.");
        return symbolName;
    }

    public getEnvironmentSymbolOrThrow(): string {
        const symbolName = this.registry.getSymbolNameById(this.getEnvironmentSymbolId());
        assertDefined(symbolName, "Environment symbol not found.");
        return symbolName;
    }

    public getRequestsContainerSymbolOrThrow(): string {
        const symbolName = this.registry.getSymbolNameById(this.getRequestsContainerSymbolId());
        assertDefined(symbolName, `Requests container symbol not found`);
        return symbolName;
    }

    /**
     * Retrieves the registered fully qualified inline request type symbol name for a given endpoint and request.
     */
    public getFullyQualifiedRequestTypeSymbolOrThrow(endpointId: string, requestNamePascalCase: string): string {
        const symbolName = this.getRequestTypeSymbolOrThrow(endpointId, requestNamePascalCase);
        const containerSymbolName = this.getRequestsContainerSymbolOrThrow();
        return `${containerSymbolName}.${symbolName}`;
    }

    public getRequestTypeSymbolOrThrow(endpointId: string, requestNamePascalCase: string): string {
        const symbolId = this.getRequestTypeSymbolId(endpointId, requestNamePascalCase);
        const symbolName = this.requestsRegistry.getSymbolNameById(symbolId);
        assertDefined(symbolName, `Request symbol not found for request '${requestNamePascalCase}'`);
        return symbolName;
    }

    public getAllRequestTypeSymbols() {
        return this.requestsRegistry.getAllSymbols();
    }

    /**
     * Retrieves the registered sub-client symbol name for a given subpackage.
     *
     * @param subpackageId The unique identifier of the subpackage
     * @returns The sub-client symbol name for the specified subpackage
     * @throws Error if no sub-client symbol has been registered for the subpackage
     */
    public getSubClientSymbolOrThrow(subpackageId: string): string {
        const symbolName = this.registry.getSymbolNameById(this.getSubClientSymbolId(subpackageId));
        assertDefined(symbolName, `Subclient symbol not found for subpackage ${subpackageId}`);
        return symbolName;
    }

    /**
     * Retrieves the registered schema type symbol name for a given type ID.
     *
     * @param typeId The unique identifier of the schema type
     * @returns The schema type symbol name for the specified type
     * @throws Error if no schema type symbol has been registered for the type ID
     */
    public getSchemaTypeSymbolOrThrow(typeId: string): string {
        const symbolName = this.registry.getSymbolNameById(this.getSchemaTypeSymbolId(typeId));
        assertDefined(symbolName, `Schema type symbol not found for type ${typeId}`);
        return symbolName;
    }

    /**
     * Registers a unique symbol name for the module.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The registered unique module symbol name
     */
    public registerModuleSymbol({
        configModuleName,
        apiNamePascalCase
    }: {
        configModuleName: string | undefined;
        apiNamePascalCase: string;
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}`,
            `${apiNamePascalCase}Api`,
            `${apiNamePascalCase}Module`
        ];
        if (typeof configModuleName === "string") {
            candidates.unshift(configModuleName);
        }
        return this.registry.registerSymbol(this.getModuleSymbolId(), candidates);
    }

    /**
     * Registers a unique symbol name for the root client class.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The registered unique root client symbol name
     */
    public registerRootClientSymbol({
        configClientClassName,
        apiNamePascalCase
    }: {
        configClientClassName: string | undefined;
        apiNamePascalCase: string;
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}Client`,
            `${apiNamePascalCase}ApiClient`,
            `${apiNamePascalCase}Service`
        ];
        if (typeof configClientClassName === "string") {
            candidates.unshift(configClientClassName);
        }
        return this.registry.registerSymbol(this.getRootClientSymbolId(), candidates);
    }

    /**
     * Registers and generates a unique symbol name for the environment enum.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The generated unique environment symbol name
     */
    public registerEnvironmentSymbol({
        configEnvironmentEnumName,
        apiNamePascalCase
    }: {
        configEnvironmentEnumName: string | undefined;
        apiNamePascalCase: string;
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}Environment`,
            `${apiNamePascalCase}Environ`,
            `${apiNamePascalCase}Env`
        ];
        if (typeof configEnvironmentEnumName === "string") {
            candidates.unshift(configEnvironmentEnumName);
        }
        return this.registry.registerSymbol(this.getEnvironmentSymbolId(), candidates);
    }

    public registerRequestsContainerSymbol(): string {
        return this.registry.registerSymbol(this.getRequestsContainerSymbolId(), [
            "Requests",
            "RequestTypes",
            "InlineRequests"
        ]);
    }

    /**
     * Registers and generates a unique symbol name for an inline request type.
     * Generates different fallback candidates based on whether the request name already ends with "Request".
     *
     * @param endpointId The unique identifier of the endpoint
     * @param requestNamePascalCase The request name in PascalCase
     * @returns The generated unique inline request type symbol name
     */
    public registerRequestTypeSymbol({
        endpointId,
        requestNamePascalCase
    }: {
        endpointId: string;
        requestNamePascalCase: string;
    }): string {
        const symbolId = this.getRequestTypeSymbolId(endpointId, requestNamePascalCase);
        const fallbackCandidates: string[] = [`${requestNamePascalCase}Type`];
        if (requestNamePascalCase.endsWith("Request")) {
            fallbackCandidates.push(`${requestNamePascalCase}Body`, `${requestNamePascalCase}BodyType`);
        } else {
            fallbackCandidates.push(
                `${requestNamePascalCase}Request`,
                `${requestNamePascalCase}RequestBody`,
                `${requestNamePascalCase}RequestBodyType`
            );
        }
        return this.requestsRegistry.registerSymbol(symbolId, [requestNamePascalCase, ...fallbackCandidates]);
    }

    /**
     * Registers and generates a unique symbol name for a sub-client class.
     * Generates fallback candidates by combining filepath parts with the subpackage name
     * to create unique identifiers when simple names collide.
     *
     * @param subpackageId The unique identifier of the subpackage
     * @param fernFilepathPartNamesPascalCase Array of filepath parts in PascalCase
     * @param subpackageNamePascalCase The subpackage name in PascalCase
     * @returns The generated unique sub-client symbol name
     */
    public registerSubClientSymbol({
        subpackageId,
        fernFilepathPartNamesPascalCase,
        subpackageNamePascalCase
    }: {
        subpackageId: string;
        fernFilepathPartNamesPascalCase: string[];
        subpackageNamePascalCase: string;
    }): string {
        const reversedParts = fernFilepathPartNamesPascalCase.toReversed();
        reversedParts.shift();
        const fallbackCandidates = reversedParts.map(
            (_, partIdx) =>
                reversedParts
                    .slice(0, partIdx + 1)
                    .reverse()
                    .join("") +
                subpackageNamePascalCase +
                "Client"
        );
        return this.registry.registerSymbol(this.getSubClientSymbolId(subpackageId), [
            `${subpackageNamePascalCase}Client`,
            ...fallbackCandidates
        ]);
    }

    /**
     * Registers and generates a unique symbol name for a schema type (struct, enum, union, etc.).
     * Tries candidate names in order: {Type}, {Type}Type, {Type}Model, {Type}Schema.
     *
     * @param typeId The unique identifier of the type
     * @param typeDeclarationNamePascalCase The type declaration name in PascalCase
     * @returns The generated unique schema type symbol name
     */
    public registerSchemaTypeSymbol(typeId: string, typeDeclarationNamePascalCase: string): string {
        return this.registry.registerSymbol(this.getSchemaTypeSymbolId(typeId), [
            typeDeclarationNamePascalCase,
            `${typeDeclarationNamePascalCase}Type`,
            `${typeDeclarationNamePascalCase}Model`,
            `${typeDeclarationNamePascalCase}Schema`
        ]);
    }

    private getModuleSymbolId(): string {
        return `${SYMBOL_ID_PREFIX}module`;
    }

    private getRootClientSymbolId(): string {
        return `${SYMBOL_ID_PREFIX}root_client_class`;
    }

    private getEnvironmentSymbolId(): string {
        return `${SYMBOL_ID_PREFIX}environment_enum`;
    }

    private getRequestsContainerSymbolId(): string {
        return `${SYMBOL_ID_PREFIX}requests_container`;
    }

    private getSubClientSymbolId(subpackageId: string): string {
        return `${SYMBOL_ID_PREFIX}subpackage_client_${subpackageId}`;
    }

    private getSchemaTypeSymbolId(typeId: string): string {
        return `${SYMBOL_ID_PREFIX}schema_type_${typeId}`;
    }

    /**
     * @returns The symbol ID to use for the requests registry.
     */
    private getRequestTypeSymbolId(endpointId: string, requestNamePascalCase: string): string {
        return `${endpointId}_${requestNamePascalCase}`;
    }
}
