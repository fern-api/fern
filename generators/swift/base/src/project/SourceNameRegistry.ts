import { SymbolRegistry as Namespace } from "@fern-api/core-utils";
import { TargetSymbolRegistry } from "./target-symbol-registry";

class ModuleNamespace {
    private readonly namespace: Namespace;

    public constructor() {
        this.namespace = new Namespace();
    }

    // Name IDs

    private asIsNameId(symbolName: string): string {
        return `AsIs:${symbolName}`;
    }

    private rootClientNameId(): string {
        return `RootClientClass`;
    }

    private environmentNameId(): string {
        return `EnvironmentEnum`;
    }

    private requestsContainerNameId(): string {
        return `RequestsContainer`;
    }

    private requestTypeNameId(endpointId: string, requestNamePascalCase: string): string {
        return `RequestType:${endpointId}:${requestNamePascalCase}`;
    }

    private subClientNameId(subpackageId: string): string {
        return `SubClient:${subpackageId}`;
    }

    // Getters

    public getRootClientNameOrThrow() {
        return this.namespace.getSymbolNameByIdOrThrow(this.rootClientNameId());
    }

    public getEnvironmentNameOrThrow() {
        return this.namespace.getSymbolNameByIdOrThrow(this.environmentNameId());
    }

    public getRequestsContainerNameOrThrow() {
        return this.namespace.getSymbolNameByIdOrThrow(this.requestsContainerNameId());
    }

    public getRequestTypeNameOrThrow(endpointId: string, requestNamePascalCase: string) {
        return this.namespace.getSymbolNameByIdOrThrow(this.requestTypeNameId(endpointId, requestNamePascalCase));
    }

    public getSubClientNameOrThrow(subpackageId: string) {
        return this.namespace.getSymbolNameByIdOrThrow(this.subClientNameId(subpackageId));
    }

    // Setters

    public addAsIsSymbol(symbolName: string) {
        const nameId = this.asIsNameId(symbolName);
        this.namespace.registerSymbol(nameId, [symbolName]);
    }

    public addRootClientSymbol(symbolNameCandidates: [string, ...string[]]) {
        const nameId = this.rootClientNameId();
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }

    public addEnvironmentSymbol(symbolNameCandidates: [string, ...string[]]) {
        const nameId = this.environmentNameId();
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }

    public addRequestsContainerSymbol(symbolNameCandidates: [string, ...string[]]) {
        const nameId = `RequestsContainer`;
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }

    public addRequestTypeSymbol(endpointId: string, requestNamePascalCase: string) {
        const nameId = this.requestTypeNameId(endpointId, requestNamePascalCase);
        return this.namespace.registerSymbol(nameId, [requestNamePascalCase]);
    }

    public addSubClientSymbol(subpackageId: string, symbolNameCandidates: [string, ...string[]]) {
        const nameId = this.subClientNameId(subpackageId);
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }

    public addSchemaTypeSymbol(typeId: string, symbolNameCandidates: [string, ...string[]]) {
        const nameId = `SchemaType:${typeId}`;
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }
}

class RequestsNamespace {
    private readonly namespace: Namespace;

    public constructor() {
        this.namespace = new Namespace();
    }

    public addRequestTypeSymbol(
        endpointId: string,
        requestNamePascalCase: string,
        symbolNameCandidates: [string, ...string[]]
    ) {
        const nameId = `${endpointId}_${requestNamePascalCase}`;
        return this.namespace.registerSymbol(nameId, symbolNameCandidates);
    }
}

export class SourceNameRegistry {
    public static create(): SourceNameRegistry {
        return new SourceNameRegistry();
    }

    private readonly targetSymbolRegistry: TargetSymbolRegistry;
    private readonly moduleNamespace: ModuleNamespace;
    private readonly requestsNamespace: RequestsNamespace;

    private constructor() {
        this.targetSymbolRegistry = TargetSymbolRegistry.create();
        this.moduleNamespace = new ModuleNamespace();
        this.requestsNamespace = new RequestsNamespace();
    }

    public getModuleNameOrThrow() {
        return this.targetSymbolRegistry.getModuleSymbolOrThrow();
    }

    public getRootClientNameOrThrow() {
        return this.moduleNamespace.getRootClientNameOrThrow();
    }

    public getEnvironmentNameOrThrow() {
        return this.moduleNamespace.getEnvironmentNameOrThrow();
    }

    public getRequestsContainerNameOrThrow() {
        return this.moduleNamespace.getRequestsContainerNameOrThrow();
    }

    public getRequestTypeNameOrThrow(endpointId: string, requestNamePascalCase: string) {
        return this.moduleNamespace.getRequestTypeNameOrThrow(endpointId, requestNamePascalCase);
    }

    public getSubClientNameOrThrow(subpackageId: string) {
        return this.moduleNamespace.getSubClientNameOrThrow(subpackageId);
    }

    /**
     * Registers a unique symbol name for the module.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The registered module symbol ID.
     */
    public registerModuleSymbol({
        configModuleName,
        apiNamePascalCase,
        asIsSymbolNames
    }: {
        configModuleName: string | undefined;
        apiNamePascalCase: string;
        asIsSymbolNames: string[];
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}`,
            `${apiNamePascalCase}Api`,
            `${apiNamePascalCase}Module`
        ];
        if (typeof configModuleName === "string") {
            candidates.unshift(configModuleName);
        }
        const tempNamespace = new Namespace({ reservedSymbolNames: ["Swift", "Foundation"] });
        const moduleSymbolName = tempNamespace.registerSymbol("Module", candidates);
        const moduleSymbolId = this.targetSymbolRegistry.registerModule(moduleSymbolName);
        asIsSymbolNames.forEach((asIsSymbolName) => {
            this.targetSymbolRegistry.registerType(asIsSymbolName);
            this.moduleNamespace.addAsIsSymbol(asIsSymbolName);
        });
        return moduleSymbolId;
    }

    /**
     * Registers a unique symbol name for the root client class.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The registered root client symbol ID.
     */
    public registerRootClientSymbol({
        configClientClassName,
        apiNamePascalCase
    }: {
        configClientClassName: string | undefined;
        apiNamePascalCase: string;
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}Client`,
            `${apiNamePascalCase}ApiClient`,
            `${apiNamePascalCase}Service`
        ];
        if (typeof configClientClassName === "string") {
            candidates.unshift(configClientClassName);
        }
        const symbolName = this.moduleNamespace.addRootClientSymbol(candidates);
        return this.targetSymbolRegistry.registerType(symbolName);
    }

    /**
     * Registers and generates a unique symbol name for the environment enum.
     * Tries preferred name first, then falls back to standard candidates.
     *
     * @returns The generated environment symbol ID.
     */
    public registerEnvironmentSymbol({
        configEnvironmentEnumName,
        apiNamePascalCase
    }: {
        configEnvironmentEnumName: string | undefined;
        apiNamePascalCase: string;
    }): string {
        const candidates: [string, ...string[]] = [
            `${apiNamePascalCase}Environment`,
            `${apiNamePascalCase}Environ`,
            `${apiNamePascalCase}Env`
        ];
        if (typeof configEnvironmentEnumName === "string") {
            candidates.unshift(configEnvironmentEnumName);
        }
        const symbolName = this.moduleNamespace.addEnvironmentSymbol(candidates);
        return this.targetSymbolRegistry.registerType(symbolName);
    }

    public registerRequestsContainerSymbol(): string {
        const symbolName = this.moduleNamespace.addRequestsContainerSymbol([
            "Requests",
            "RequestTypes",
            "InlineRequests"
        ]);
        return this.targetSymbolRegistry.registerType(symbolName);
    }

    /**
     * Registers and generates a unique symbol name for an inline request type.
     * Generates different fallback candidates based on whether the request name already ends with "Request".
     *
     * @param endpointId The unique identifier of the endpoint
     * @param requestNamePascalCase The request name in PascalCase
     * @returns The generated unique inline request type symbol ID.
     */
    public registerRequestTypeSymbol({
        endpointId,
        requestNamePascalCase
    }: {
        endpointId: string;
        requestNamePascalCase: string;
    }): string {
        const fallbackCandidates: string[] = [`${requestNamePascalCase}Type`];
        if (requestNamePascalCase.endsWith("Request")) {
            fallbackCandidates.push(`${requestNamePascalCase}Body`, `${requestNamePascalCase}BodyType`);
        } else {
            fallbackCandidates.push(
                `${requestNamePascalCase}Request`,
                `${requestNamePascalCase}RequestBody`,
                `${requestNamePascalCase}RequestBodyType`
            );
        }
        const symbolName = this.requestsNamespace.addRequestTypeSymbol(endpointId, requestNamePascalCase, [
            requestNamePascalCase,
            ...fallbackCandidates
        ]);
        return this.targetSymbolRegistry.registerType(symbolName);
    }

    /**
     * Registers and generates a unique symbol name for a sub-client class.
     * Generates fallback candidates by combining filepath parts with the subpackage name
     * to create unique identifiers when simple names collide.
     *
     * @param subpackageId The unique identifier of the subpackage
     * @param fernFilepathPartNamesPascalCase Array of filepath parts in PascalCase
     * @param subpackageNamePascalCase The subpackage name in PascalCase
     * @returns The generated unique sub-client symbol ID.
     */
    public registerSubClientSymbol({
        subpackageId,
        fernFilepathPartNamesPascalCase,
        subpackageNamePascalCase
    }: {
        subpackageId: string;
        fernFilepathPartNamesPascalCase: string[];
        subpackageNamePascalCase: string;
    }): string {
        const reversedParts = fernFilepathPartNamesPascalCase.toReversed();
        reversedParts.shift();
        const fallbackCandidates = reversedParts.map(
            (_, partIdx) =>
                reversedParts
                    .slice(0, partIdx + 1)
                    .reverse()
                    .join("") +
                subpackageNamePascalCase +
                "Client"
        );
        const symbolName = this.moduleNamespace.addSubClientSymbol(subpackageId, [
            `${subpackageNamePascalCase}Client`,
            ...fallbackCandidates
        ]);
        return this.targetSymbolRegistry.registerType(symbolName);
    }

    /**
     * Registers and generates a unique symbol name for a schema type (struct, enum, union, etc.).
     * Tries candidate names in order: {Type}, {Type}Type, {Type}Model, {Type}Schema.
     *
     * @param typeId The unique identifier of the type
     * @param typeDeclarationNamePascalCase The type declaration name in PascalCase
     * @returns The generated unique schema type symbol ID.
     */
    public registerSchemaTypeSymbol(typeId: string, typeDeclarationNamePascalCase: string): string {
        const symbolName = this.moduleNamespace.addSchemaTypeSymbol(typeId, [
            typeDeclarationNamePascalCase,
            `${typeDeclarationNamePascalCase}Type`,
            `${typeDeclarationNamePascalCase}Model`,
            `${typeDeclarationNamePascalCase}Schema`
        ]);
        return this.targetSymbolRegistry.registerType(symbolName);
    }
}
