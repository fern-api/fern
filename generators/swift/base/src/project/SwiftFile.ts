import { File } from "@fern-api/base-generator";
import { RelativeFilePath } from "@fern-api/fs-utils";

export declare namespace SwiftFile {
    interface Args {
        filename: string;
        directory: RelativeFilePath;
        fileContents: string;
    }
}

export class SwiftFile extends File {
    constructor(args: SwiftFile.Args) {
        super(args.filename, args.directory, args.fileContents);
    }
}
