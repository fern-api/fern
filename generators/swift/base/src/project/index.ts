export * from "./SwiftFile";
export * from "./SwiftProject";
