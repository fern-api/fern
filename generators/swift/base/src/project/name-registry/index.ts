export * from "./source-name-registry";
