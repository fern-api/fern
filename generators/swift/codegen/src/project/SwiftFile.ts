import { File } from "@fern-api/base-generator";
import { RelativeFilePath } from "@fern-api/fs-utils";

import { Enum } from "../ast/Enum";
import { Struct } from "../ast/Struct";

export declare namespace SwiftFile {
    interface Args {
        /* The class to be written to the CSharp File */
        clazz: Struct | Enum;
        /* Directory of the filepath */
        directory: RelativeFilePath;
    }
}

export class SwiftFile extends File {
    constructor({ clazz, directory }: SwiftFile.Args) {
        super(`${clazz.name}.cs`, directory, clazz.toString());
    }
}
