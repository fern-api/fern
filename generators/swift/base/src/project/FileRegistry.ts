import { RelativeFilePath } from "@fern-api/fs-utils";
import { swift } from "@fern-api/swift-codegen";
import { SwiftFile } from "./SwiftFile";

export class FileRegistry {
    private readonly filesByName: Map<string, SwiftFile> = new Map();

    public getAll(): SwiftFile[] {
        return Array.from(this.filesByName.values());
    }

    public add({
        nameCandidateWithoutExtension,
        directory,
        contents,
        includeFoundationImport
    }: {
        nameCandidateWithoutExtension: string;
        directory: RelativeFilePath;
        contents: string | swift.FileComponent[];
        includeFoundationImport?: true;
    }): SwiftFile {
        let filenameWithoutExt = nameCandidateWithoutExtension;
        while (this.filesByName.has(filenameWithoutExt)) {
            filenameWithoutExt += "_";
        }
        const filename = filenameWithoutExt + ".swift";
        const contentsArray = typeof contents === "string" ? [contents] : contents;
        const file = includeFoundationImport
            ? SwiftFile.createWithFoundation({
                  filename,
                  directory,
                  contents: contentsArray
              })
            : SwiftFile.create({
                  filename,
                  directory,
                  contents: contentsArray
              });
        this.filesByName.set(filenameWithoutExt, file);
        return file;
    }
}
