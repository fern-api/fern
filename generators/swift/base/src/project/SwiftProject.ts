import { mkdir } from "fs/promises";

import { AbstractProject } from "@fern-api/base-generator";
import { RelativeFilePath, join } from "@fern-api/fs-utils";
import { BaseSwiftCustomConfigSchema } from "@fern-api/swift-codegen";

import { AbstractSwiftGeneratorContext } from "../context/AbstractSwiftGeneratorContext";
import { SwiftFile } from "./SwiftFile";

const SRC_DIRECTORY_NAME = "src";

export class SwiftProject extends AbstractProject<AbstractSwiftGeneratorContext<BaseSwiftCustomConfigSchema>> {
    private name: string;
    private sourceFiles: SwiftFile[] = [];

    public constructor({
        context,
        name
    }: {
        context: AbstractSwiftGeneratorContext<BaseSwiftCustomConfigSchema>;
        name: string;
    }) {
        super(context);
        this.name = name;
    }

    public get sourceFileDirectory(): RelativeFilePath {
        return RelativeFilePath.of(SRC_DIRECTORY_NAME);
    }

    public get projectDirectory(): RelativeFilePath {
        return join(this.sourceFileDirectory, RelativeFilePath.of(this.name));
    }

    public addSourceFiles(...files: SwiftFile[]): void {
        this.sourceFiles.push(...files);
    }

    public async persist(): Promise<void> {
        const absolutePathToSrcDirectory = join(this.absolutePathToOutputDirectory, this.sourceFileDirectory);
        this.context.logger.debug(`mkdir ${absolutePathToSrcDirectory}`);
        await mkdir(absolutePathToSrcDirectory, { recursive: true });

        const absolutePathToProjectDirectory = join(absolutePathToSrcDirectory, RelativeFilePath.of(this.name));
        this.context.logger.debug(`mkdir ${absolutePathToProjectDirectory}`);
        await mkdir(absolutePathToProjectDirectory, { recursive: true });

        await Promise.all(this.sourceFiles.map((file) => file.write(absolutePathToProjectDirectory)));
    }
}
