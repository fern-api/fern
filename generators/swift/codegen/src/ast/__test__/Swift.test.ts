import Swift, { AccessLevel, ClassLevel, FunctionModifier } from "../..";
import { SwiftFile } from "../../project/SwiftFile";
import { VariableType } from "../VariableType";

describe("Swift Language", () => {
    it("makes file header", () => {
        const output = Swift.makeFileHeader({
            header: "This is the header to a file"
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes import", () => {
        const output = Swift.makeImport({
            packageName: "ExamplePackage"
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes enum case", () => {
        const output = Swift.makeEnumCase({
            name: "fallbackContent",
            key: "fallback_content"
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes enum", () => {
        const output = Swift.makeEnum({
            name: "CodingKeys",
            inheritance: [
                Swift.makeType({
                    name: "String"
                }),
                Swift.makeType({
                    name: "CodingKey"
                })
            ],
            enumCases: [
                Swift.makeEnumCase({
                    name: "name"
                }),
                Swift.makeEnumCase({
                    name: "fallbackContent",
                    key: "fallback_content"
                })
            ]
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes function", () => {
        const output = Swift.makeFunc({
            accessLevel: AccessLevel.Fileprivate,
            modifier: FunctionModifier.Static,
            name: "makePerson",
            async: true,
            params: [
                Swift.makeParam({
                    title: "name",
                    type: Swift.makeType({
                        name: "String"
                    })
                }),
                Swift.makeParam({
                    title: "age",
                    type: Swift.makeType({
                        name: "Int"
                    }),
                    defaultValue: Swift.makeType({
                        name: "Int"
                    })
                })
            ]
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes primitives", () => {
        const output = Swift.makeField({
            name: "count",
            valueType: Swift.makePrimitive({
                key: "integer"
            })
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes struct", () => {
        const output = Swift.makeStruct({
            accessLevel: AccessLevel.Public,
            name: "TopGun",
            inheritance: [
                Swift.makeType({
                    name: "Movie"
                })
            ],
            fields: [
                Swift.makeField({
                    name: "description",
                    variableType: VariableType.Let,
                    valueType: Swift.makePrimitive({
                        key: "string"
                    })
                })
            ]
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes type", () => {
        const output = Swift.makeType({
            accessLevel: AccessLevel.Open,
            classLevel: ClassLevel.Struct,
            name: "ExampleObject",
            functions: [
                Swift.makeFunc({
                    name: "fetchData"
                })
            ],
            inheritance: [
                Swift.makeType({
                    name: "Codable"
                }),
                Swift.makeType({
                    name: "NSObject"
                })
            ]
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });

    it("makes file", async () => {
        const output = Swift.makeFile({
            fileHeader: Swift.makeFileHeader({
                header: "// Sample.swift"
            }),
            imports: [Swift.makeImport({ packageName: "Foundation" }), Swift.makeImport({ packageName: "UIKit" })],
            node: Swift.makeType({
                accessLevel: AccessLevel.Open,
                classLevel: ClassLevel.Class,
                name: "Room",
                subclasses: [
                    Swift.makeType({
                        classLevel: ClassLevel.Class,
                        name: "Person",
                        functions: [
                            Swift.makeFunc({
                                name: "getName"
                            })
                        ]
                    }),
                    Swift.makeEnum({
                        name: "RoomType",
                        enumCases: [
                            Swift.makeEnumCase({
                                name: "big"
                            }),
                            Swift.makeEnumCase({
                                name: "small",
                                key: "sml"
                            })
                        ]
                    })
                ],
                functions: [
                    Swift.makeFunc({
                        name: "openDoor"
                    }),
                    Swift.makeFunc({
                        accessLevel: AccessLevel.Public,
                        modifier: FunctionModifier.Static,
                        name: "closeDoor",
                        async: true,
                        throws: true,
                        returnType: Swift.makeType({
                            name: "Int"
                        })
                    })
                ]
            })
        });
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        expect(output.toString()).toMatchSnapshot();
    });
});
