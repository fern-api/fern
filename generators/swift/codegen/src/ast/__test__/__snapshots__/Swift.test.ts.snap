// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Swift Language > makes enum 1`] = `"enum CodingKeys: String, CodingKey {case namecase fallbackContent = fallback_content}"`;

exports[`Swift Language > makes enum case 1`] = `"case fallbackContent = fallback_content"`;

exports[`Swift Language > makes file 1`] = `
"// Sample.swift
import Foundationimport UIKit
open class Room {
    class Person {
        func getName() {print("Hey!")}
    }
    enum RoomType {case bigcase small = sml}
    func openDoor() {print("Hey!")}
    public static func closeDoor() async throws -> Int {print("Hey!")}
}
"
`;

exports[`Swift Language > makes file header 1`] = `"This is the header to a file"`;

exports[`Swift Language > makes function 1`] = `
"fileprivate static func makePerson(name: String {
}, age: Int {
} =  FIX THIS) async {print("Hey!")}"
`;

exports[`Swift Language > makes import 1`] = `"import ExamplePackage"`;

exports[`Swift Language > makes primitives 1`] = `" let count: Int"`;

exports[`Swift Language > makes struct 1`] = `"public struct TopGun: Movie { let description: String}"`;

exports[`Swift Language > makes type 1`] = `
"open struct ExampleObject: Codable, NSObject {
    func fetchData() {print("Hey!")}
}"
`;
